### PR TITLE
[generate_financial_metrics] Write unit tests and function body

### PR DESCRIPTION
**Pull request by dorisyycai**
It's related to https://github.com/UBC-MDS/quanteda/issues/30

This commit includes:
   - function implementation
   - test functions for generate_financial_metrics

### CODE REVIEW

File: tests/test_generate_financial_metrics.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
+import pytest
+import pandas as pd
+import numpy as np
+from quanteda.generate_return_series import generate_return_series
+from quanteda.generate_financial_metrics import generate_financial_metrics
+
+idx=pd.date_range("2018-01-01", periods=5, freq="H")
```
Comment (JohnShiuMK) [inline tests/test_generate_financial_metrics.py:7]:
```suggestion
hourly_idx = pd.date_range("2018-01-01", periods=5, freq="H")
```

File: tests/test_generate_financial_metrics.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+input_df = pd.DataFrame(pd.Series(range(len(idx)), index=idx))
```
Comment (JohnShiuMK) [inline tests/test_generate_financial_metrics.py:8]:
```suggestion
test_hourly_df = pd.DataFrame(pd.Series(range(len(idx)), index=hourly_idx))
```

File: tests/test_generate_financial_metrics.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
```
Comment (JohnShiuMK) [inline tests/test_generate_financial_metrics.py:9]:
```suggestion
weekly_idx = pd.date_range("2018-01-01", periods=5, freq="W")
```

File: tests/test_generate_financial_metrics.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
```
Comment (JohnShiuMK) [inline tests/test_generate_financial_metrics.py:10]:
```suggestion
test_weekly_df = pd.DataFrame(pd.Series(range(len(idx)), index=weekly_idx))
```

File: tests/test_generate_financial_metrics.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        generate_financial_metrics(pd.DataFrame(df), annual_risk_free='not numeric')
+
+def test_invalid_frequency():
+    """
+    Checking if the input dataframe index frequency is either daily, hourly or minutely
+    """
+    with pytest.raises(ValueError, match="Invalid frequency. Frequency should be either 'D' for daily, 'H' for hourly, or 'min' for minutely."):
+        idx=pd.date_range("2018-01-01", periods=5, freq="W")
+        df = pd.DataFrame(pd.Series(range(len(idx)), index=idx))
+        generate_financial_metrics(pd.DataFrame(df))
```
Comment (JohnShiuMK) [inline tests/test_generate_financial_metrics.py:47]:
```suggestion
        generate_financial_metrics(test_weekly_df)
```

File: tests/test_generate_financial_metrics.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        df = pd.DataFrame(pd.Series(range(len(idx)), index=idx))
+        generate_financial_metrics(pd.DataFrame(df), annual_risk_free='not numeric')
+
+def test_invalid_frequency():
+    """
+    Checking if the input dataframe index frequency is either daily, hourly or minutely
+    """
+    with pytest.raises(ValueError, match="Invalid frequency. Frequency should be either 'D' for daily, 'H' for hourly, or 'min' for minutely."):
+        idx=pd.date_range("2018-01-01", periods=5, freq="W")
+        df = pd.DataFrame(pd.Series(range(len(idx)), index=idx))
```
Comment (JohnShiuMK) [inline tests/test_generate_financial_metrics.py:46]:
```suggestion
```

File: tests/test_generate_financial_metrics.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    with pytest.raises(ValueError, match="Annual risk-free rate must be numeric."):
+        df = pd.DataFrame(pd.Series(range(len(idx)), index=idx))
+        generate_financial_metrics(pd.DataFrame(df), annual_risk_free='not numeric')
+
+def test_invalid_frequency():
+    """
+    Checking if the input dataframe index frequency is either daily, hourly or minutely
+    """
+    with pytest.raises(ValueError, match="Invalid frequency. Frequency should be either 'D' for daily, 'H' for hourly, or 'min' for minutely."):
+        idx=pd.date_range("2018-01-01", periods=5, freq="W")
```
Comment (JohnShiuMK) [inline tests/test_generate_financial_metrics.py:45]:
```suggestion
```

File: tests/test_generate_financial_metrics.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    """
+    with pytest.raises(ValueError, match="Random returns DataFrame contains missing values."):
+        generate_financial_metrics(pd.DataFrame([np.nan, np.nan, 1, 2, 3], index=idx))
+
+def test_risk_free_rate_type():
+    """
+    Checking if the input annual_risk_free is numeric
+    """
+    with pytest.raises(ValueError, match="Annual risk-free rate must be numeric."):
+        df = pd.DataFrame(pd.Series(range(len(idx)), index=idx))
```
Comment (JohnShiuMK) [inline tests/test_generate_financial_metrics.py:37]:
```suggestion
```

File: tests/test_generate_financial_metrics.py [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+        generate_financial_metrics(pd.DataFrame(df), annual_risk_free='not numeric')
```
Comment (JohnShiuMK) [inline tests/test_generate_financial_metrics.py:38]:
```suggestion
        generate_financial_metrics(test_hourly_df, annual_risk_free='not numeric')
```

File: src/quanteda/generate_financial_metrics.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+    ----------
+    random_returns_df : pd.DataFrame
+        DataFrame containing random return values with a datetime index.
+
+    annual_risk_free : float, default 0.00.
+        Annualized risk-free rate as a decimal (e.g., 0.02 for 2%).
+        
+    Returns
+    -------
+    pd.DataFrame:
```
Comment (JohnShiuMK) [inline src/quanteda/generate_financial_metrics.py:18]:
```suggestion
    DataFrame
```

File: src/quanteda/generate_financial_metrics.py [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+        DataFrame containing financial metrics including count (number of returns),
```
Comment (JohnShiuMK) [inline src/quanteda/generate_financial_metrics.py:19]:
```suggestion
        A DataFrame containing financial metrics including count (number of returns), 
```

File: src/quanteda/generate_financial_metrics.py [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+        DataFrame containing random return values with a datetime index.
+
+    annual_risk_free : float, default 0.00.
+        Annualized risk-free rate as a decimal (e.g., 0.02 for 2%).
+        
+    Returns
+    -------
+    pd.DataFrame:
+        DataFrame containing financial metrics including count (number of returns), 
+        total return, annual return, annual volatility, and Sharpe ratio.
```
Comment (JohnShiuMK) [inline src/quanteda/generate_financial_metrics.py:20]:
```suggestion
        total return, annualized return, annualized volatility, and Sharpe ratio.
```

**Review by JohnShiuMK — CHANGES_REQUESTED** [review]

File: tests/test_generate_financial_metrics.py (hunk 1 above)
Comment (dorisyycai) [reply tests/test_generate_financial_metrics.py]: good suggestion! Pls see my latest update.

File: tests/test_generate_financial_metrics.py (hunk 2 above)
Comment (dorisyycai) [reply tests/test_generate_financial_metrics.py]: good suggestion! Pls see my latest update.

File: tests/test_generate_financial_metrics.py [hunk 13]
```diff
@@ -0,0 +1,61 @@
+import pytest
+import pandas as pd
+import numpy as np
+from quanteda.generate_return_series import generate_return_series
+from quanteda.generate_financial_metrics import generate_financial_metrics
+
+idx_h=pd.date_range("2018-01-01", periods=5, freq="H")
```
Comment (JohnShiuMK) [inline tests/test_generate_financial_metrics.py:7]:
```suggestion
idx_h = pd.date_range("2018-01-01", periods=5, freq="H")
```

File: tests/test_generate_financial_metrics.py [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+idx_w=pd.date_range("2018-01-01", periods=5, freq="W")
```
Comment (JohnShiuMK) [inline tests/test_generate_financial_metrics.py:8]:
```suggestion
idx_w = pd.date_range("2018-01-01", periods=5, freq="W")
```

**Review by JohnShiuMK — CHANGES_REQUESTED** [review]

**Review by JohnShiuMK — APPROVED** [review]

**Comment by JohnShiuMK** [comment]
good job, thanks @dorisyycai

**Pull request merged**